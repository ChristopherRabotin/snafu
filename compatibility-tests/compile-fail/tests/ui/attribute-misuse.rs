mod enum_misuse {
    use snafu::prelude::*;

    #[derive(Debug, Snafu)]
    #[snafu(display("display should not work here"))]
    #[snafu(source(from(XXXX, Box::new)))]
    #[snafu(source(true))]
    #[snafu(backtrace)]
    #[snafu(context)]
    #[snafu(implicit)]
    #[snafu(provide)]
    #[snafu(provide(u8 => 0))]
    enum EnumError {
        AVariant,
    }
}

mod variant_misuse {
    use snafu::prelude::*;

    #[derive(Debug, Snafu)]
    enum EnumError {
        // Make sure we catch the error in the second attribute
        #[snafu(display("an error variant"), source(from(XXXX, Box::new)))]
        #[snafu(source)]
        #[snafu(backtrace)]
        #[snafu(crate_root(XXXX))]
        #[snafu(implicit)]
        #[snafu(provide)]
        AVariant,
    }
}

mod field_misuse {
    use snafu::prelude::*;

    #[derive(Debug, Snafu)]
    enum EnumError {
        AVariant {
            #[snafu(display("display should not work here"))]
            #[snafu(visibility(pub))]
            #[snafu(source(false))]
            #[snafu(source(from(XXXX, Box::new)))]
            #[snafu(context)]
            #[snafu(crate_root(XXXX))]
            source: String,

            #[snafu(provide(false))]
            #[snafu(provide(u8 => 0))]
            not_a_source: bool,
        },
    }
}

mod struct_misuse {
    use snafu::prelude::*;

    #[derive(Debug, Snafu)]
    enum UsableError {}

    #[derive(Debug, Snafu)]
    #[snafu(display("display should not work here"))]
    #[snafu(source(from(UsableError, Box::new)))]
    #[snafu(visibility(pub))]
    #[snafu(source(true))]
    #[snafu(backtrace)]
    #[snafu(context)]
    #[snafu(implicit)]
    #[snafu(provide)]
    #[snafu(provide(u8 => 0))]
    struct StructError(Box<UsableError>);
}

fn main() {}
