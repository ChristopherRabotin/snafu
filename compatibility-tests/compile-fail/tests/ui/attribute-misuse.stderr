error: `display` attribute is only valid on enum variants or structs with named fields, not on an enum
 --> $DIR/attribute-misuse.rs:5:13
  |
5 |     #[snafu(display("display should not work here"))]
  |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: `source(from)` attribute is only valid on enum variant or struct fields with a name, not on an enum
 --> $DIR/attribute-misuse.rs:6:13
  |
6 |     #[snafu(source(from(XXXX, Box::new)))]
  |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: `source(bool)` attribute is only valid on enum variant or struct fields with a name, not on an enum
 --> $DIR/attribute-misuse.rs:7:13
  |
7 |     #[snafu(source(true))]
  |             ^^^^^^^^^^^^

error: `backtrace` attribute is only valid on enum variant or struct fields with a name, not on an enum
 --> $DIR/attribute-misuse.rs:8:13
  |
8 |     #[snafu(backtrace)]
  |             ^^^^^^^^^

error: `context(bool)` attribute is only valid on enum variants or structs with named fields, not on an enum
 --> $DIR/attribute-misuse.rs:9:13
  |
9 |     #[snafu(context)]
  |             ^^^^^^^

error: `implicit` attribute is only valid on enum variant or struct fields with a name, not on an enum
  --> $DIR/attribute-misuse.rs:10:13
   |
10 |     #[snafu(implicit)]
   |             ^^^^^^^^

error: `source` attribute is only valid on enum variant or struct fields with a name, not on an enum variant
  --> $DIR/attribute-misuse.rs:22:46
   |
22 |         #[snafu(display("an error variant"), source(from(XXXX, Box::new)))]
   |                                              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: `source` attribute is only valid on enum variant or struct fields with a name, not on an enum variant
  --> $DIR/attribute-misuse.rs:23:17
   |
23 |         #[snafu(source)]
   |                 ^^^^^^

error: `backtrace` attribute is only valid on enum variant or struct fields with a name, not on an enum variant
  --> $DIR/attribute-misuse.rs:24:17
   |
24 |         #[snafu(backtrace)]
   |                 ^^^^^^^^^

error: `crate_root` attribute is only valid on an enum or a struct, not on an enum variant
  --> $DIR/attribute-misuse.rs:25:17
   |
25 |         #[snafu(crate_root(XXXX))]
   |                 ^^^^^^^^^^^^^^^^

error: `implicit` attribute is only valid on enum variant or struct fields with a name, not on an enum variant
  --> $DIR/attribute-misuse.rs:26:17
   |
26 |         #[snafu(implicit)]
   |                 ^^^^^^^^

error: `display` attribute is only valid on enum variants or structs with named fields, not on a field
  --> $DIR/attribute-misuse.rs:37:21
   |
37 |             #[snafu(display("display should not work here"))]
   |                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: `visibility` attribute is only valid on an enum, enum variants, or a struct with named fields, not on a field
  --> $DIR/attribute-misuse.rs:38:21
   |
38 |             #[snafu(visibility(pub))]
   |                     ^^^^^^^^^^^^^^^

error: Incompatible attributes [`source(false)`, `source(from)`] specified on a field
  --> $DIR/attribute-misuse.rs:40:21
   |
40 |             #[snafu(source(from(XXXX, Box::new)))]
   |                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: `context` attribute is only valid on enum variants or structs with named fields, not on a field
  --> $DIR/attribute-misuse.rs:41:21
   |
41 |             #[snafu(context)]
   |                     ^^^^^^^

error: `crate_root` attribute is only valid on an enum or a struct, not on a field
  --> $DIR/attribute-misuse.rs:42:21
   |
42 |             #[snafu(crate_root(XXXX))]
   |                     ^^^^^^^^^^^^^^^^

error: `display` attribute is only valid on enum variants or structs with named fields, not on a tuple struct
  --> $DIR/attribute-misuse.rs:55:13
   |
55 |     #[snafu(display("display should not work here"))]
   |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: `visibility` attribute is only valid on an enum, enum variants, or a struct with named fields, not on a tuple struct
  --> $DIR/attribute-misuse.rs:57:13
   |
57 |     #[snafu(visibility(pub))]
   |             ^^^^^^^^^^^^^^^

error: `source(bool)` attribute is only valid on enum variant or struct fields with a name, not on a tuple struct
  --> $DIR/attribute-misuse.rs:58:13
   |
58 |     #[snafu(source(true))]
   |             ^^^^^^^^^^^^

error: `backtrace` attribute is only valid on enum variant or struct fields with a name, not on a tuple struct
  --> $DIR/attribute-misuse.rs:59:13
   |
59 |     #[snafu(backtrace)]
   |             ^^^^^^^^^

error: `context` attribute is only valid on enum variants or structs with named fields, not on a tuple struct
  --> $DIR/attribute-misuse.rs:60:13
   |
60 |     #[snafu(context)]
   |             ^^^^^^^

error: `implicit` attribute is only valid on enum variant or struct fields with a name, not on a tuple struct
  --> $DIR/attribute-misuse.rs:61:13
   |
61 |     #[snafu(implicit)]
   |             ^^^^^^^^
